doc/arch/007-screen-architecture-v2.md

```diff
@@ -0,0 +1,80 @@
+- `UiState`: This is a Kotlin data class that represents everything needed to render the content of a given screen.
+- `Ui`: This is an interface which represents all the functionality that the actual screen needs to provide to the controller.
+- `UiChange`: This is a Kotlin lambda that has the signature `(Ui) -> Unit`.
+
+#### Screen Setup Process
+- Define a `UiState` for whatever view/screen is being built.
+- Create a `UiStateProducer` which is an `ObservableTransformer<Event, UiState>`. This class will be responsible for business logic by transforming a stream of events into a stream of UI states.
+- Create a `UiChangeProducer` which is an `ObservableTransformer<UiState, UiChange>`. This is responsible for presentation logic by transforming a stream of UI states into a stream of lambdas which will executed on the `Ui`.
+- Create a controller which is an `ObservableTransformer<UiEvent, UiChange>` which will compose `UiStateProducer` and `UiChangeProducer` internally.
+- Use the `ViewControllerBinding` to tie the event stream to the view.
+
+##### Sample
+```kotlin
+@Inject
+lateinit var uiStateProducer: ShortCodeSearchResultStateProducer
+
+@Inject
+lateinit var uiChangeProducer: ShortCodeSearchResultUiChangeProducer
+
+lateinit var binding: ViewControllerBinding<UiEvent, ShortCodeSearchResultState, ShortCodeSearchResultUi>
+
+override fun onFinishInflate() {
+	// Inject the screen
+
+  binding = ViewControllerBinding.bindToView(this, uiStateProducer, uiChangeProducer)
+
+  newPatientButton.setOnClickListener { binding.onEvent(SearchPatient) }
+}
+```
+
+### Goals Review
+
+#### Separation of concerns and testability
+Since the business logic and presentation logic are separated into two discrete components, testing them is a lot simpler since they can be tested independently of each other. In addition, we can test the behaviour of the business logic using value testing (asserting the generated states) which is a lot more readable than verifying behavious using mocks.
+
+#### UI state saving/restoration
+Since the state of a screen is represented as a data class, saving and restoring the state is quite easy, and just has two steps:
+1. Make the `UiState` class implement the `Parcelable` interface.
+2. The `ViewControllerBinding` exposes two functions, `latestState()` and `restoreSavedState()`, which can be used in conjunction with the platform lifecycle methods.
+
+#### Migrating from the older architecture
+One of the major concerns about moving from the older architecture to the newer one is breaking existing features while moving code around. Since both the older architecture and new expose the same overall interface (`ObservableTransformer<Event, UiChange>`), we can replace the controller in tests for the older screens with a composition of the state producer and state consumer and verify that no behaviour was changed when migrating.
```

Comment:
Let's specify which architecture we're specifically talking about instead of saying "newer one"? Say:
> One of the major concerns about moving from the older architecture to the ~newer one~ **v2 architecture** is breaking existing features while moving code around

Reply:
Done.